### PR TITLE
Enable the creation of concurrent, but not connected, output and previous connectors

### LINKED ISSUE #5691: Enable having both previous and output connections on a block
**Is your feature request related to a problem? Please describe.**

I would like to be able to create a "chameleon" block which would have previous, next and output connections when
created, so that it could be used as either a value or a statement.  Once it is actually connected it would change to an appropriate shape depending on where its context.

**Describe the solution you'd like**

- I'd like to be able to use `Block.setOutput(true)` when the block has a previous (but not connected) connection. 
- I'd like to be able to use `Block.setPreviousConnectio(true)` when the block has an (unconnected) output connection.
- I'd like the connection checker to enforce the restriction on not having connected previous and output connections.

**Describe alternatives you've considered**

Having a less dynamic "chameleon" block which was only a statement or a value but which had a context menu item (or other UI affordance) to change its shape.

**Additional context**

**Note that I intend to submit a PR to implement this feature request.**

See [this thread](https://groups.google.com/g/blockly/c/mmZcSqoGun8) and [this one ](https://groups.google.com/g/blockly/c/QPvDH2cj7-g/)in the Blockly forum for some discussion on this issue.

### PR DESCRIPTION
**Pull request by mark-friedman**
## The basics

- [x] I branched from develop
- [x] My pull request is against develop
- [x] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)

### Resolves

Resolves #5691

### Proposed Changes

- Enables the creation of concurrent output and previous connections via `Block.setOutput(true)` and `Block.setPreviousConnection(true)`
- Moves enforcement of the behavior to the `ConnectionChecker` to ensure that even though output and previous connectors are allowed on a block, actual concurrent connections to other blocks cannot be made.
- Adds tests for the new `ConnectionChecker` behavior.
- Modifies `Block.unplug` to no longer assume that a Block will have either a previous or an output connection but not both.

#### Behavior Before Change

- `Block.setOutput(true)` would throw an error if the block already had a previous connection
- `Block.setPreviousConnection(true)` would throw an error if the block already had an output connection
- `Block.unplug` assumed that a Block will have either a previous or an output connection but not both.

#### Behavior After Change

Please see the "Proposed Changes" section.

### Reason for Changes

I would like to be able to create a "chameleon" block which would have previous, next and output connections when
created, so that it could be used as either a value or a statement. Once it is actually connected it would change to an appropriate shape depending on its context.  I would imagine that there are other, similar, use cases.

PR #5691 has some more context and links to some further discussion in the Blockly forum.

### Test Coverage

Added unit tests to connection_checker_test.js.

* Desktop Chrome

### Documentation

I'm not sure that it says anywhere in the documentation that a block cannot have both a previous and an output connection, but if there is it should be changed to reflect this PR.

### Additional Information

It's quite possible that there is code other than `Block.unplug` that assumes that Blocks only have either an output or a previous connection.  Running all the tests did not uncover any such an assumptions, but they may still exist. If so, I would be happy to update this PR.

I rewrote `ConnectionChecker.doSafetyChecks` to use more explanatory variable names (in addition to adding my new checks).  I found the original code a bit difficult to follow and hope that the new variable names make it a little clearer.  If you don't like my changes, I can change the code to be as before, but still with my new checks.

### CODE REVIEW

**Comment by mark-friedman** [comment]
@googlebot I signed it!

**Comment by rachel-fenichel** [comment]
> It's quite possible that there is code other than Block.unplug that assumes that Blocks only have either an output or a previous connection. Running all the tests did not uncover any such an assumptions, but they may still exist. If so, I would be happy to update this PR.

Did you test in the playground by dragging around to connect and disconnect blocks? Mid-drag behaviour is not well tested by the mocha tests.

**Comment by mark-friedman** [comment]
> Did you test in the playground by dragging around to connect and disconnect blocks? Mid-drag behaviour is not well tested by the mocha tests.

Yes, I did a pretty decent amount of manual testing in a playground.

**Comment by mark-friedman** [comment]
I realized that I had based my change on an older version of the develop branch, so I merged in the latest changes to develop, made some necessary code changes and retested.

**Comment by mark-friedman** [comment]
I realized that my develop branch was out of date, so I pulled down the latest from the Blockly repo and new testing showed some more places that had assumptions that a block could have only one of an output and a previous connector.  My latest commit, above, deals with those pieces of code.

**Comment by mark-friedman** [comment]
Just a note that this PR and #5722 are somewhat of a team in that they are both necessary for my goal of creating a dynamic "chameleon" block. That said, strictly speaking they are independent.

**Comment by rachel-fenichel** [comment]
Tests pass, except for lint and clang-format, which should be trivially fixable.

I pulled this down and dragged around for a bit and it looks good. I'll approve once you fix formatting and rename the helper function I commented on.

**Comment by mark-friedman** [comment]
> I'll approve once you fix formatting and rename the helper function I commented on.

@rachel-fenichel, Thanks!  I see the lint formatting comments, but I don't see comments about any helper functions.

File: core/keyboard_nav/ast_node.js [hunk 1]
```diff
@@ -231,6 +231,25 @@ ASTNode.createWorkspaceNode = function(workspace, wsCoordinate) {
...
+ * Gets the top connection on a block.
+ * This is either an output connection, previous connection or undefined.
+ * If both connections exist return the one that is actually connected
+ * to another block.
+ * @param {!Block} block The block to find the top connection on.
+ * @return {Connection} The connection holding the top most position on the
+ *     block.
+ * @private
+ */
+ASTNode.getTopConnection_ = function(block) {
```
[the remark this answers is not in the document]
Comment (mark-friedman) [reply core/keyboard_nav/ast_node.js:244]: Sure.  I started with something like your suggestion but then thought that maybe I should try to be consistent with the rest of the code.

"A foolish consistency is the hobgoblin of small minds" - Ralph Waldo Emerson

[the remark this answers is not in the document]
Comment (rachel-fenichel) [reply core/keyboard_nav/ast_node.js:244]: No worries--we generally go for consistency, but also we rewrote all of our code this quarter to get to `goog.module`, so all the old rules are out the door.

**Comment by mark-friedman** [comment]
PTAL

File: core/keyboard_nav/ast_node.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (mark-friedman) [reply core/keyboard_nav/ast_node.js]: Understood.

**Comment by rachel-fenichel** [comment]
Clang-format is still failing--run `npm run format` to fix.

(eslint can tell you when your lines are too long, but clang-format intelligently breaks lines up)

**Comment by mark-friedman** [comment]
OK. Let's try again!

**Review by rachel-fenichel — APPROVED** [review]
Hooray! Thanks for your code.

**Comment by mark-friedman** [comment]
Woohoo!  Thanks again, Rachel!

-Mark

**Pull request merged**